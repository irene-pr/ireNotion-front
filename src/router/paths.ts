const paths = {
  home: "/",
  login: "/login",
  register: "/register",
  userBoard: "/userBoard",
};

export default paths;
