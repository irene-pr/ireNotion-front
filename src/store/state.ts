const state = {
  isLoggedIn: false,
  userContent: {
    name: "",
    boards: [],
  },
  userData: {
    userId: "",
    username: "",
  },
  themeHeaders: "night-mode",
  themeSurfaces: "day-mode",
  isBoardEditModal: false,
  isUpdateNoteModal: false,
  idForModal: "",
};

export default state;
