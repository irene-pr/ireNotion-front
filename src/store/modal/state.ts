import { IModalState } from "@/types/store";

const state = (): IModalState => ({
  isBoardEditModal: false,
  isUpdateNoteModal: false,
  idForModal: "",
});

export default state;
