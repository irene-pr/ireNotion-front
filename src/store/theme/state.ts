import { IThemeState } from "@/types/store";

const state = (): IThemeState => ({
  themeHeaders: "night-mode",
  themeSurfaces: "day-mode",
});

export default state;
