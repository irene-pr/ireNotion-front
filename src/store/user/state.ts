import { IUserState } from "@/types/store";

const state = (): IUserState => ({
  isLoggedIn: false,
  userContent: {
    name: "",
    boards: [],
  },
  userData: {
    userId: "",
    username: "",
  },
  isLoading: false,
});

export default state;
