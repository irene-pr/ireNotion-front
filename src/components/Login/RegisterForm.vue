<template>
  <div class="register-container">
    <router-link :to="paths.home">
      <h1 class="register-container__main-heading">ire<span>N</span>otion</h1>
    </router-link>

    <form class="register-form" @submit.prevent="onSubmit" autocomplete="off">
      <div class="register-form__input-container">
        <label class="register-form__label" for="name">Name</label>
        <input
          type="text"
          id="name"
          class="register-form__input"
          v-model="name"
          placeholder="Mario"
        />
        <label class="register-form__label" for="username">Username</label>
        <input
          type="text"
          id="username"
          class="register-form__input"
          v-model="username"
          placeholder="mario123"
        />
        <label class="register-form__label" for="password">Password</label>
        <input
          type="password"
          id="password"
          class="register-form__input"
          v-model="password"
          placeholder="at least 7 characters"
        />
        <label class="register-form__label" for="repeat-password"
          >Repeat password</label
        >
        <input
          type="password"
          id="repeat-password"
          class="register-form__input"
          v-model="repeatPassword"
          placeholder="at least 7 characters"
        />
      </div>
      <p class="message">
        {{ messageShown }}
      </p>
      <button
        class="register-form__button register-form__button--signup"
        type="submit"
        :disabled="
          name === '' ||
          username === '' ||
          password === '' ||
          repeatPassword === ''
        "
      >
        SIGN <span>UP</span>
      </button>
    </form>

    <div class="divider">
      <div class="divider__line divider__line--first"></div>
      <p class="divider__or">or</p>
      <div class="divider__line divider__line--second"></div>
    </div>
    <router-link :to="paths.login">
      <button class="register-form__button register-form__button--login">
        LOG <span>IN</span>
      </button>
    </router-link>
  </div>
</template>

<script lang="ts">
import { defineComponent } from "vue";
import { mapActions } from "vuex";
import { IUserRegisterData } from "@/types/interfaces";
import paths from "../../router/paths";
import router from "@/router";

export default defineComponent({
  name: "RegisterForm",
  components: {},
  data() {
    return {
      paths,
      name: "",
      username: "",
      password: "",
      repeatPassword: "",
      messageShown: "",
    };
  },
  methods: {
    ...mapActions(["registerUser"]),
    async onSubmit() {
      if (
        this.name !== "" &&
        this.username !== "" &&
        this.password !== "" &&
        this.repeatPassword !== "" &&
        this.password === this.repeatPassword &&
        this.password.length >= 7 &&
        this.password.length <= 20
      ) {
        const userRegisterData: IUserRegisterData = {
          name: this.name,
          username: this.username,
          password: this.password,
        };

        await this.registerUser(userRegisterData);
        router.push(paths.login);
      } else if (this.password !== this.repeatPassword) {
        this.messageShown = "The passwords don't match. Try again";
      } else if (this.password.length < 7 || this.password.length > 20) {
        this.messageShown =
          "The password must have between 7 and 20 characters";
      }
    },
  },
});
</script>

<style lang="scss" scoped>
@import "@/styles/_variables";
@import "@/styles/_extends";

.register-container {
  width: 380px;
  padding-bottom: 20px;

  display: flex;
  flex-direction: column;
  align-items: center;
  background-color: $theme-light-color;
  color: $theme-dark-color;
  font-family: $Lato;

  a {
    text-decoration: none;
  }
  &__main-heading {
    font-family: $h1;
    font-size: 40px;
    font-weight: 100;
    color: $theme-dark-color;
    span {
      color: $theme-blue;
    }
  }
  &__secondary-heading {
    color: $theme-dark-color-opacity;
    font-family: $h2;
    font-size: 20px;
    font-weight: 500;
    letter-spacing: 0.03em;
  }
  @media only screen and (max-width: 480px) {
    height: 100%;
    width: 100%;
    justify-content: center;
  }
}
.register-form {
  padding: 30px 0 10px 0;
  display: flex;
  flex-direction: column;
  align-items: center;

  &__input-container {
    display: flex;
    flex-direction: column;
  }

  &__label {
    padding: 5px 0 5px 2px;
    @extend %label-text;
  }
  &__input {
    height: 30px;
    padding-bottom: 0;
    border: none;
    border-bottom: 1px solid $theme-form-line;
    background-color: transparent;
    color: $theme-dark-color;
    font-size: 14px;
    &::placeholder {
      font-family: $Lato;
      opacity: 0.85;
      letter-spacing: 0;
    }
    &[type="password"] {
      font-family: Verdana, serif;
      letter-spacing: 0.125em;
    }
    &:focus {
      outline: none;
      color: $theme-blue;
    }
  }
  &__button {
    @extend %button;
    position: initial;
    &--login {
      @extend %button--login;
      margin: 12px 0;
    }
    &--signup {
      @extend %button--signup;
    }
    &:disabled {
      opacity: 0.6;
      cursor: initial;
    }
    span {
      @extend %button--span;
      color: $theme-blue;
    }
  }
}
.message {
  height: 15px;
  margin: 20px 0;
  @extend %label-text;
  font-size: 12px;
  color: $theme-blue;
}
.divider {
  width: 40%;
  display: flex;
  justify-content: space-around;
  align-items: center;
  &__or {
    @extend %label-text;
  }
  &__line {
    width: 50px;
    height: 0.1px;
    border-bottom: 1px solid $theme-form-line;
  }
}
</style>
