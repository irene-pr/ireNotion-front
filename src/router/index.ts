import { createRouter, createWebHistory, RouteRecordRaw } from "vue-router";
import paths from "@/router/paths";
import Home from "../views/Home.vue";
import Login from "../views/Login.vue";
import Register from "../views/Register.vue";
import BoardPage from "../views/BoardPage.vue";

const routes: Array<RouteRecordRaw> = [
  {
    path: paths.home,
    name: "Home",
    component: Home,
  },
  {
    path: paths.login,
    name: "Login",
    component: Login,
  },
  {
    path: paths.register,
    name: "Register",
    component: Register,
  },
  {
    path: paths.userBoard,
    name: "BoardPage",
    component: BoardPage,
  },
];

const router = createRouter({
  history: createWebHistory(process.env.BASE_URL),
  routes,
});

export default router;
