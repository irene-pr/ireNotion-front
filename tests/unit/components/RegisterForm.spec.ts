import { createStore } from "vuex";
import { mount } from "@vue/test-utils";
import state from "@/store/state";
import RegisterForm from "@/components/Login/RegisterForm.vue";
import router from "@/router";

describe("Given a RegisterForm component", () => {
  let store = createStore({
    state() {
      return state;
    },
  });

  describe("When it is rendered", () => {
    test("Then it renders", () => {
      mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });
    });
    test("Then the it matches the snapshot", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });
      expect(wrapper.element).toMatchSnapshot();
    });
    test("Then it renders a main header", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const mainHeader = wrapper.find("h1");
      expect(mainHeader.text()).toContain("ireNotion");
    });

    test("Then it renders the login button", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const button = wrapper.find("button.register-form__button--login");

      expect(button.text()).toContain("LOG IN");
    });
    test("Then the divider renders an 'or", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const divider = wrapper.find("div.divider");

      expect(divider.html()).toContain('<p class="divider__or">or</p>');
    });
    test("Then it renders a form", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const form = wrapper.findAll("form");

      expect(form).toHaveLength(1);
    });
  });
  describe("When it renders the form", () => {
    test("Then the form contains an input and a label for the name", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const form = wrapper.find("form");
      const nameInput = form.find('input[id="name"]');
      const nameLabel = form.find('label[for="name"]');

      expect(nameInput.exists).toBeTruthy();
      expect(nameLabel.exists).toBeTruthy();
    });
    test("Then the form contains an input and a label for the username", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const form = wrapper.find("form");
      const usernameInput = form.find('input[id="username"]');
      const usernameLabel = form.find('label[for="username"]');

      expect(usernameInput.exists).toBeTruthy();
      expect(usernameLabel.exists).toBeTruthy();
    });
    test("Then the form contains an input and a label for the password", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const form = wrapper.find("form");
      const passwordInput = form.find('input[id="password"]');
      const passwordLabel = form.find('label[for="password"]');

      expect(passwordInput.exists).toBeTruthy();
      expect(passwordLabel.exists).toBeTruthy();
    });
    test("Then the form contains a signup button", () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const button = wrapper.find("button.register-form__button--signup");

      expect(button.text()).toContain("SIGN UP");
    });
  });

  describe("When I don't write on the inputs", () => {
    test("Then the signup button must be disabled", async () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const button = wrapper.find("button.register-form__button--signup");

      expect(button.element.getAttributeNames()).toContain("disabled");
    });
    test("The message is not triggered", async () => {
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      await wrapper
        .find("button.register-form__button--signup")
        .trigger("click");

      expect(wrapper.find(".message").text()).toBe("");
    });
  });
  describe("When I write on the inputs", () => {
    test("Then the value is set", async () => {
      store = createStore({
        state() {
          return state;
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue("password");
      inputRepeatPassword.setValue("repeatPassword");

      expect(inputName.element.value).toBe("name");
      expect(inputUsername.element.value).toBe("username");
      expect(inputPassword.element.value).toBe("password");
      expect(inputRepeatPassword.element.value).toBe("repeatPassword");
    });
  });
  describe("When I write different passwords on the inputs and I click the signup button", () => {
    test("Then the message 'The passwords don't match. Try again' is shown", async () => {
      store = createStore({
        state() {
          return state;
        },
        actions: {
          registerUser: jest.fn().mockResolvedValue(401),
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue("password");
      inputRepeatPassword.setValue("repeatPassword");

      await wrapper.find("form").trigger("submit.prevent");

      expect(wrapper.vm.$data.messageShown).toBe(
        "The passwords don't match. Try again"
      );
    });
  });
  describe("When I write matching passwords with a length lesser than 7 on the inputs and I click the signup button", () => {
    test("Then the message 'The password must have between 7 and 20 characters' is shown", async () => {
      store = createStore({
        state() {
          return state;
        },
        actions: {
          registerUser: jest.fn().mockResolvedValue(401),
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue("pass");
      inputRepeatPassword.setValue("pass");

      await wrapper.find("form").trigger("submit.prevent");

      expect(wrapper.vm.$data.messageShown).toBe(
        "The password must have between 7 and 20 characters"
      );
    });
  });
  describe("When I write matching passwords with a length greater than 20 on the inputs and I click the signup button", () => {
    test("Then the message 'The password must have between 7 and 20 characters' is shown", async () => {
      store = createStore({
        state() {
          return state;
        },
        actions: {
          registerUser: jest.fn().mockResolvedValue(401),
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        stubs: ["router-view"],
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue(
        "passssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss"
      );
      inputRepeatPassword.setValue(
        "passssssssssssssssssssssssssssssssssssssssssssssssssssssssssssss"
      );

      await wrapper.find("form").trigger("submit.prevent");

      expect(wrapper.vm.$data.messageShown).toBe(
        "The password must have between 7 and 20 characters"
      );
    });
  });
  describe("When I write an existing username in the input and I click the signup button", () => {
    test("Then the message reads 'Username already exists'", async () => {
      const $route = { path: "/userBoard" };
      store = createStore({
        state() {
          return state;
        },
        actions: {
          registerUser: jest.fn().mockResolvedValue("Username already exists"),
        },
        getters: {
          redirectToLogin: jest.fn(),
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        mocks: {
          $route,
        },
        stubs: ["router-view", "router-link"],
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue("password");
      inputRepeatPassword.setValue("password");

      await wrapper.find("form").trigger("submit.prevent");

      expect(wrapper.vm.$data.messageShown).toBe("Username already exists");
    });
  });
  describe("When I write right data on the inputs and I click the signup button", () => {
    test("Then it the message remains unchanged", async () => {
      const onSubmitMock = jest.fn();
      const $route = { path: "/userBoard" };
      store = createStore({
        state() {
          return state;
        },
        actions: {
          registerUser: jest.fn(),
        },
        getters: {
          redirectToLogin: jest.fn(),
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        mocks: {
          $route,
        },
        stubs: ["router-view", "router-link"],
        methods: {
          onSubmit: onSubmitMock,
        },
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue("password");
      inputRepeatPassword.setValue("password");

      await wrapper.find("form").trigger("submit.prevent");

      expect(onSubmitMock).toHaveBeenCalled();
    });
    test("Then it the message remains unchanged", async () => {
      const redirectToLoginMock = jest.fn();
      const $route = { path: "/userBoard" };
      store = createStore({
        state() {
          return state;
        },
        actions: {
          registerUser: jest.fn().mockResolvedValue(200),
        },
        getters: {
          redirectToLogin: redirectToLoginMock,
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        mocks: {
          $route,
        },
        stubs: ["router-view", "router-link"],
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue("password");
      inputRepeatPassword.setValue("password");

      await wrapper.find("form").trigger("submit.prevent");

      expect(redirectToLoginMock).toHaveBeenCalled();
    });
    test("Then it the message remains unchanged", async () => {
      const $route = { path: "/userBoard" };
      store = createStore({
        state() {
          return state;
        },
        actions: {
          registerUser: jest.fn().mockResolvedValue(200),
        },
        getters: {
          redirectToUserBoard: jest.fn(),
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        mocks: {
          $route,
        },
        stubs: ["router-view", "router-link"],
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue("password");
      inputRepeatPassword.setValue("password");

      expect(wrapper.find(".message").text()).toBe("");
    });
    test("Then it the message data value remains unchanged", async () => {
      const $route = { path: "/userBoard" };
      store = createStore({
        state() {
          return state;
        },
        actions: {
          registerUser: jest.fn().mockResolvedValue(200),
        },
        getters: {
          redirectToUserBoard: jest.fn(),
        },
      });
      const wrapper = mount(RegisterForm, {
        global: {
          plugins: [router, store],
        },
        mocks: {
          $route,
        },
        stubs: ["router-view", "router-link"],
      });

      const [inputName, inputUsername, inputPassword, inputRepeatPassword] =
        wrapper.findAll("input");
      inputName.setValue("name");
      inputUsername.setValue("username");
      inputPassword.setValue("password");
      inputRepeatPassword.setValue("password");

      expect(wrapper.vm.$data.messageShown).toBe("");
    });
  });
});
