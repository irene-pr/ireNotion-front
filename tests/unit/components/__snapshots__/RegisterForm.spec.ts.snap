// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Given a RegisterForm component When it is rendered Then the it matches the snapshot 1`] = `
<div
  class="register-container"
>
  <a
    class=""
    href="/"
  >
    <h1
      class="register-container__main-heading"
    >
      ire
      <span>
        N
      </span>
      otion
    </h1>
  </a>
  <form
    autocomplete="off"
    class="register-form"
  >
    <div
      class="register-form__input-container"
    >
      <label
        class="register-form__label"
        for="name"
      >
        Name
      </label>
      <input
        class="register-form__input"
        id="name"
        placeholder="Mario"
        type="text"
      />
      <label
        class="register-form__label"
        for="username"
      >
        Username
      </label>
      <input
        class="register-form__input"
        id="username"
        placeholder="mario123"
        type="text"
      />
      <label
        class="register-form__label"
        for="password"
      >
        Password
      </label>
      <input
        class="register-form__input"
        id="password"
        placeholder="at least 7 characters"
        type="password"
      />
      <label
        class="register-form__label"
        for="repeat-password"
      >
        Repeat password
      </label>
      <input
        class="register-form__input"
        id="repeat-password"
        placeholder="at least 7 characters"
        type="password"
      />
    </div>
    <p
      class="message"
    />
    <button
      class="register-form__button register-form__button--signup"
      disabled=""
      type="submit"
    >
       SIGN 
      <span>
        UP
      </span>
    </button>
  </form>
  <div
    class="divider"
  >
    <div
      class="divider__line divider__line--first"
    />
    <p
      class="divider__or"
    >
      or
    </p>
    <div
      class="divider__line divider__line--second"
    />
  </div>
  <a
    class=""
    href="/login"
  >
    <button
      class="register-form__button register-form__button--login"
    >
       LOG 
      <span>
        IN
      </span>
    </button>
  </a>
</div>
`;
